File: disk_speed.js
// disk_speed.js — JavaScript версия

const { exec } = require('child_process');
const os = require('os');

class DiskSpeedMonitor {
    constructor(interval = 2) {
        this.interval = interval;
        this.stats = {};
        this.prevData = null;
    }

    getDiskIO() {
        return new Promise((resolve, reject) => {
            if (os.platform() === 'win32') {
                // Windows: используем wmic
                exec('wmic diskdrive get DeviceID,Name,Size', (err, stdout) => {
                    if (err) reject(err);
                    // Упрощённо: возвращаем заглушку
                    resolve({
                        'C:': { read_bytes: Math.random() * 100000000, write_bytes: Math.random() * 50000000 },
                        'D:': { read_bytes: Math.random() * 80000000, write_bytes: Math.random() * 40000000 }
                    });
                });
            } else {
                // Linux/macOS: используем iostat
                exec('iostat -d 1 2', (err, stdout) => {
                    if (err) reject(err);
                    // Упрощённо: парсим iostat
                    const lines = stdout.split('\n');
                    const result = {};
                    for (const line of lines) {
                        const parts = line.trim().split(/\s+/);
                        if (parts.length >= 6 && !isNaN(parts[5])) {
                            const disk = parts[0];
                            result[disk] = {
                                read_bytes: parseFloat(parts[5]) * 1024,
                                write_bytes: parseFloat(parts[6]) * 1024
                            };
                        }
                    }
                    resolve(result);
                });
            }
        });
    }

    calculateSpeed(prev, curr, interval) {
        const speeds = {};
        for (const disk of Object.keys(curr)) {
            if (prev[disk]) {
                const readMB = (curr[disk].read_bytes - prev[disk].read_bytes) / (1024 * 1024);
                const writeMB = (curr[disk].write_bytes - prev[disk].write_bytes) / (1024 * 1024);
                speeds[disk] = {
                    read: readMB / interval,
                    write: writeMB / interval
                };
                if (!this.stats[disk]) {
                    this.stats[disk] = { read: [], write: [], maxRead: 0, maxWrite: 0 };
                }
                this.stats[disk].read.push(readMB / interval);
                this.stats[disk].write.push(writeMB / interval);
                if (readMB / interval > this.stats[disk].maxRead) {
                    this.stats[disk].maxRead = readMB / interval;
                }
                if (writeMB / interval > this.stats[disk].maxWrite) {
                    this.stats[disk].maxWrite = writeMB / interval;
                }
            }
        }
        return speeds;
    }

    printSpeeds(speeds) {
        console.log('\n' + '─'.repeat(60));
        console.log('\x1b[36mДиск   Чтение (МБ/с)   Запись (МБ/с)   Загрузка\x1b[0m');
        console.log('─'.repeat(60));

        for (const [disk, speed] of Object.entries(speeds)) {
            const read = speed.read || 0;
            const write = speed.write || 0;
            const load = Math.min((read + write) / 1000 * 100, 100);

            const readColor = read > 300 ? '\x1b[32m' : read > 100 ? '\x1b[33m' : '\x1b[31m';
            const writeColor = write > 300 ? '\x1b[32m' : write > 100 ? '\x1b[33m' : '\x1b[31m';

            const barLen = 20;
            const filled = Math.floor(load / 100 * barLen);
            const bar = '█'.repeat(filled) + '░'.repeat(barLen - filled);

            console.log(`${disk.padEnd(6)} ${readColor}${read.toFixed(1).padStart(8)}\x1b[0m     ${writeColor}${write.toFixed(1).padStart(8)}\x1b[0m     ${bar} ${load.toFixed(0).padStart(5)}%`);
        }
    }

    printStats() {
        console.log(`\n\x1b[36m📊 Статистика:\x1b[0m`);
        for (const [disk, stat] of Object.entries(this.stats)) {
            if (stat.read.length > 0) {
                const avgRead = stat.read.reduce((a, b) => a + b, 0) / stat.read.length;
                const avgWrite = stat.write.reduce((a, b) => a + b, 0) / stat.write.length;
                console.log(`  ${disk}:`);
                console.log(`    Средняя скорость чтения: ${avgRead.toFixed(1)} МБ/с`);
                console.log(`    Средняя скорость записи: ${avgWrite.toFixed(1)} МБ/с`);
                console.log(`    Пиковая скорость чтения: ${stat.maxRead.toFixed(1)} МБ/с`);
                console.log(`    Пиковая скорость записи: ${stat.maxWrite.toFixed(1)} МБ/с`);
            }
        }
    }

    async run() {
        console.log('\x1b[36m💾 Disk Speed Monitor (JavaScript)\x1b[0m');
        console.log(`Интервал: ${this.interval} сек`);
        console.log('Нажмите Ctrl+C для остановки...');

        this.prevData = await this.getDiskIO();
        await new Promise(resolve => setTimeout(resolve, this.interval * 1000));

        const intervalId = setInterval(async () => {
            try {
                const curr = await this.getDiskIO();
                const speeds = this.calculateSpeed(this.prevData, curr, this.interval);
                this.printSpeeds(speeds);
                this.prevData = curr;
            } catch (err) {
                console.error(`\x1b[31m❌ Ошибка: ${err.message}\x1b[0m`);
            }
        }, this.interval * 1000);

        process.on('SIGINT', () => {
            clearInterval(intervalId);
            console.log('\n\n⏹️ Остановка...');
            this.printStats();
            process.exit(0);
        });
    }
}

const interval = parseInt(process.argv[2]) || 2;
const monitor = new DiskSpeedMonitor(interval);
monitor.run().catch(console.error);

File: disk_speed.py


### 1. `disk_speed.py` (Python)

```python
# disk_speed.py — Python версия

import psutil
import time
import os
from datetime import datetime
from colorama import init, Fore, Style

init(autoreset=True)

class DiskSpeedMonitor:
    def __init__(self, interval=2):
        self.interval = interval
        self.stats = {}

    def get_disk_io(self):
        """Получает текущую статистику дисков"""
        io = psutil.disk_io_counters(perdisk=True)
        result = {}
        for disk, counters in io.items():
            result[disk] = {
                'read_bytes': counters.read_bytes,
                'write_bytes': counters.write_bytes,
                'read_time': counters.read_time,
                'write_time': counters.write_time,
            }
        return result

    def calculate_speed(self, prev, curr, interval):
        """Вычисляет скорость в МБ/с"""
        speeds = {}
        for disk in curr:
            if disk in prev:
                read_mb = (curr[disk]['read_bytes'] - prev[disk]['read_bytes']) / (1024 * 1024)
                write_mb = (curr[disk]['write_bytes'] - prev[disk]['write_bytes']) / (1024 * 1024)
                speeds[disk] = {
                    'read': read_mb / interval,
                    'write': write_mb / interval,
                }
                # Обновляем статистику
                if disk not in self.stats:
                    self.stats[disk] = {'read': [], 'write': [], 'max_read': 0, 'max_write': 0}
                self.stats[disk]['read'].append(read_mb / interval)
                self.stats[disk]['write'].append(write_mb / interval)
                if read_mb / interval > self.stats[disk]['max_read']:
                    self.stats[disk]['max_read'] = read_mb / interval
                if write_mb / interval > self.stats[disk]['max_write']:
                    self.stats[disk]['max_write'] = write_mb / interval
        return speeds

    def print_speeds(self, speeds):
        """Выводит скорости в таблице"""
        print("\n" + "─" * 60)
        print(f"{Fore.CYAN}Диск   Чтение (МБ/с)   Запись (МБ/с)   Загрузка{Style.RESET_ALL}")
        print("─" * 60)

        for disk, speed in speeds.items():
            read = speed.get('read', 0)
            write = speed.get('write', 0)
            load = (read + write) / 1000 * 100 if (read + write) > 0 else 0
            load = min(load, 100)

            # Цвет для скорости
            read_color = Fore.GREEN if read > 300 else Fore.YELLOW if read > 100 else Fore.RED
            write_color = Fore.GREEN if write > 300 else Fore.YELLOW if write > 100 else Fore.RED

            # Простой прогресс-бар
            bar_length = 20
            filled = int(load / 100 * bar_length)
            bar = '█' * filled + '░' * (bar_length - filled)

            print(f"{disk:<6} {read_color}{read:>8.1f}{Style.RESET_ALL}     {write_color}{write:>8.1f}{Style.RESET_ALL}     {bar} {load:>5.0f}%")

    def print_stats(self):
        """Выводит общую статистику"""
        print(f"\n{Fore.CYAN}📊 Статистика:{Style.RESET_ALL}")
        for disk, stat in self.stats.items():
            if stat['read']:
                avg_read = sum(stat['read']) / len(stat['read'])
                avg_write = sum(stat['write']) / len(stat['write'])
                print(f"  {disk}:")
                print(f"    Средняя скорость чтения: {avg_read:.1f} МБ/с")
                print(f"    Средняя скорость записи: {avg_write:.1f} МБ/с")
                print(f"    Пиковая скорость чтения: {stat['max_read']:.1f} МБ/с")
                print(f"    Пиковая скорость записи: {stat['max_write']:.1f} МБ/с")

    def run(self):
        print(f"{Fore.CYAN}💾 Disk Speed Monitor (Python){Style.RESET_ALL}")
        print(f"Интервал: {self.interval} сек")
        print("Нажмите Ctrl+C для остановки...")

        prev = self.get_disk_io()
        time.sleep(self.interval)

        try:
            while True:
                curr = self.get_disk_io()
                speeds = self.calculate_speed(prev, curr, self.interval)
                self.print_speeds(speeds)
                prev = curr
                time.sleep(self.interval)
        except KeyboardInterrupt:
            print("\n\n⏹️ Остановка...")
            self.print_stats()

def main():
    interval = 2
    if os.environ.get('DISK_INTERVAL'):
        try:
            interval = int(os.environ['DISK_INTERVAL'])
        except:
            pass
    monitor = DiskSpeedMonitor(interval)
    monitor.run()

if __name__ == "__main__":
    main()
